### ISSUE TITLE
Security concern (SQL Injection)

### ISSUE THREAD

**Issue by barbenezra** [react-querybuilder/react-querybuilder#540]
I have created custom UI to create filters query (will later leverage the full potential with combinators and nesting)
![image](https://github.com/react-querybuilder/react-querybuilder/assets/5148328/b9c5c63c-e5ba-4f01-9ed6-ee35300eda97)

Wondering whether there's a possible SQL injection breach:
- I manipulate the field SQL generated from the query builder component using `transformQuery`, replacing it with values from DB
- I format the query using `formatQuery` which parseNumbers flag turned on, not using the parameterized option as it doesn't play out well with Prisma

The output of such results in:
![image](https://github.com/react-querybuilder/react-querybuilder/assets/5148328/35053fb6-ed2e-416e-b3ce-93bf468766e5)

Do you see any possible issues I should be concerned with?

**Comment by zachblume** [comment]
Definitely a problem

(1) If the UI is rendering to SQL on the frontend and then being sent to the API to be executed directly, then any user could open the devtools console and call the API themselves with an arbitrary SQL string. You should instead pass the rules object and compile to SQL on the backend.

(2) If you're already doing something like that, then the second part of protection is parameterization. 

Here is OWASP:

[https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)

> Parameterized queries force the developer to first define all the SQL code, and then pass in each parameter to the query later. This coding style allows the database to distinguish between code and data, regardless of what user input is supplied.
> 
> Prepared statements ensure that an attacker is not able to change the intent of a query, even if SQL commands are inserted by an attacker. 

"Option 1: Use of Prepared Statements (with Parameterized Queries)"

You can make prisma do it via $rawQuery:

https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access

`In turn, Prisma creates prepared statements that are safe from SQL injections:`

The problem you may face is the behavior of parameterization in this library -- e.g., does it escape fieldnames as well as value? I think it maybe not, which leave you open to injection.

(3) Parameterization is the agreed-upon standard for protecting against SQL injection, but if you choose to go down another route, you can escape every user input. So you could iterate through the value of every field and escape quotes, but you're also going to have to do that for operators, and fieldnames, and basically the value of any key output by formatQuery.

Bottom line: building an extensible SQL generator is a big attack surface and you should document and then test whatever path you go down to make sure injection is not possible. I.e., try injection yourself on your own UI and see what SQL commands get run!

**Comment by jakeboone02** [comment]
I agree completely with @zachblume here. (Thanks for the great answer, Zach!) I would definitely recommend passing the query object over the wire, not the SQL or even the parameterized SQL. Better to use `formatQuery(query, 'parameterized')` (or `'parameterized_named'`) along with any post-processing, on the server and not the client.

I investigated Prisma in the context of RQB a little bit when responding to #414 and came to the same conclusions.

**Comment by zachblume** [comment]
It might be worth looking further into this (parsing RQB to popular ORMs) and adding documentation (if not a full built-in parser) for * ease of use and * sql injection concerns. I just built a parser in an application from RQB -> postgREST query object which let me compile on the client-side, I'm sure that this is a very common use of RQB

**Comment by barbenezra** [comment]
Thanks for your detailed answers!
Just to clarify, I run the ```formatQuery``` as well as the ```transformQuery``` on the backend (not sure it's the intended behavior, but it works).

This is our UI:
![image](https://github.com/react-querybuilder/react-querybuilder/assets/5148328/2c755cf5-bad7-436d-be4e-8310bf569e2d)

![image](https://github.com/react-querybuilder/react-querybuilder/assets/5148328/862d8e05-6966-44ef-a028-d188b16b281f)

The SQL is generated on the backend, given all the inputs you select on the 'Data' panel.
I was trying to play around with different inputs and could not get around values being surrounded single quotes.

Do you see any problem with this approach? Do we still need to use parameterized query on the backend? If yes, can you give an example of why

Thanks again!

**Comment by zachblume** [comment]
[https://codesandbox.io/p/sandbox/peaceful-firefly-crk7w3?file=%2Fsrc%2FApp.tsx%3A11%2C38](https://codesandbox.io/p/sandbox/peaceful-firefly-crk7w3?file=%2Fsrc%2FApp.tsx%3A11%2C38)

Minimal reproduction of a user editing the html/request with devconsole that would induce your backend to execute an unsafe query. I injected ";BEGIN; DROP important_table; COMMIT;" as an react-querybuilder operator

You'll need to whitelist everyone operator and filedname in addition to safe-escaping the values.

Hard to patch the library for this use-case. Could wrap every field name in double quotes and safe-escape the input, but operators must be whitelisted since they're not escapable user input. The parameterized version of this suffers the same thing. I suppose you could hardcode into the library a whitelist of every possible SQL operator, but I think that may have downsides that I haven't thought up yet.

The safer way to do this is to manually compile to your ORM's spec.

**Comment by jakeboone02** [comment]
@zachblume I'm curious what you meant by "the name thing." Does the "parameterized_named" format help with that at all?

**Comment by zachblume** [comment]
"the same thing"! Sorry. Any version of this is going to face the same problem unless you whitelist operators because operators are not escapable and (in SQL's mind) they aren't user inputs in terms of paramerization. They are always going to be injectable (unless you can find a way to paramterize operators as opposed to literals -- I did a quick stack overflow search and didn't find anything!)

**Comment by pp6757185** [comment]
> 

I encountered the same issue with single quotes when using formatQuery (query, 'SQL')；How did you solve it, except for generating SQL on the backend

**Comment by barbenezra** [comment]
I ended up whitelisting possible operators sent from the client.
Since the values are being sanitized it should work fine.

Thanks
